### PR TITLE
feature: server: add support for unix socket; feature: server: refactor arg parsing; feature: ci/cd: add support for `yarn`

### PR DESCRIPTION
**Pull request by oakaigh**
- [x] commit message named according to [Contributing Guide](https://github.com/coderaiser/cloudcmd/blob/master/CONTRIBUTING.md "Contributting Guide")
- [x] `npm run fix:lint` is OK
- [x] `npm test` is OK

This PR contains multiple features:
- [x] server: support for unix socket through CLI option `--listen`
- [x] server: refactored CLI arg parsing logic for enhanced readability
- [ ] ci/cd: support for `yarn`

### CODE REVIEW

File: json/help.json
Comment (coderaiser) [inline json/help.json:1]: Why this file removed?

**Review by coderaiser — COMMENTED** [review]
Hi! Thanks for a PR!

Could you please split to couple PR's, as I see there is couple sorts of changes, let’s discuss:
- `yargs-parser` switched to `yargs`: why?
- Unix socket is good idea, but how do you use it?
- How yarn is better then npm for current repo?

**Comment by oakaigh** [comment]
To address your questions:
> `yargs-parser` switched to `yargs`: why?

Convenience. With `yargs` I was able to organize the arguments better.
I found the old way of manually generating help messages quite error-prone and confusing.
In `yargs` they are automatically generated for you.
This also explains why [json/help.json](https://github.com/coderaiser/cloudcmd/pull/401/files/e54305b8fed2f337915470bebfe265f195397815#diff-2e608469f82f0531f04f061661b1e63c6052e5077407801a6c8c2eae75ae62b9) is no longer needed - the help messages are declared alongside the arguments.

> Unix socket is good idea, but how do you use it?

A good use case would be to run multiple instances of `cloudcmd` servers behind an authenticating proxy 
(e.g. Jupyter server proxy https://github.com/jupyterhub/jupyter-server-proxy),
where each instance is spawned under a different user.
As UNIX domain sockets have builtin access control measures 
(through user permissions: perfect for a multi-user server hub),
they are inherently much more secure and reliable for IPC than TCP/IP protocols,
which by default relies on upper layers for access control.

> How yarn is better then npm for current repo?

Well, `npm` does not seem to work for this repo on my machine. 
`yarn` is the only one that I was able to use to build the project - it's certainly not a requirement though.
I guess the lack of documentation on how users should set up the development environment is the reason.
Personally I think both `yarn` and `npm` should be supported for convenience.

**Comment by coderaiser** [comment]
When you switched to `yargs` is any output of Cloud Commander changed? Or it remain the same?

File: README.md [hunk 1]
```diff
@@ -20,30 +20,42 @@
...
+```sh
 npm i cloudcmd -g
 ```
 
-## Start
+- OR from source
+
+```sh
+npm_config_build_from_source=true npm i coderaiser/cloudcmd
+```
```
Comment (coderaiser) [inline README.md:35]: Is it `npm` feature? Could you please point documentation

**Comment by oakaigh** [comment]
> When you switched to yargs is any output of Cloud Commander changed? Or it remain the same?

It definitely changed. The default values are now added to the output so users know what they are dealing with. The arguments prefixed with `no-` are dropped from the help messages (but still supported!).

But I believe the output is customizable. You are not supposed to generate the help messages manually especially when the number of arguments is HUGE - yours is close to about 50 if not more.

> yargs and Unix sockets linked together?

The CLI option `--listen` is parsed with `yargs` not `yargs-parser`. And `--listen` introduces a new addressing scheme which allows either a `<hostname>:<port>` pair or a `<path-to-unix-socket>` to be specified. Without `--listen` there would be no unix socket support, at least from the CLI.

So this entire PR is more about CLI argument parsing than the handling of UNIX sockets, as the UNIX socket support is handled by the underlying node HTTP server.

> and also docs + npm.

Apologize for the confusion. Free free to move the changes to a separate PR.

**Comment by coderaiser** [comment]
Thanks, but there is a lot of changes (even breaking), would be better if we talk about this in separate issue (with screenshots of what you suggests), also about sockets, how they should work on windows and Mac OS? All other options works good on all systems.

**Comment by oakaigh** [comment]
No problem. There aren't any "breaking" changes per se - like I said, the arguments are still supported but they are just not shown in the help message anymore.

No screenshots needed - build the project and give it a try.

> about sockets, how they should work on windows and Mac OS?

It's complicated. But UNIX domain sockets should be supported on both Windows and MacOS.
More details can be found [here](https://nodejs.org/api/net.html#serverlisten).
The use case or "how xxx works" is not a primary concern here. 
The concern is to leave users as many configurable options as possible, and leave the low level protocol handling to the node.js library.

**Comment by coderaiser** [comment]
Could you please describe use case, I don’t quite understand what is wrong with TCP

> As UNIX domain sockets have builtin access control measures
(through user permissions: perfect for a multi-user server hub),
they are inherently much more secure and reliable for IPC than TCP/IP protocols,
which by default rely on upper layers for access control.

And how interacting with Cloud Commander can be changed?
Right now default port is `8000`, and can be accessed from web browser by URL: `http://localhost:8000`, how is it changed with Unix sockets?

**Comment by oakaigh** [comment]
UNIX domain sockets are particularly useful in the IPC context.

When you use TCP in a multi user setup, without any form of authentication present, the port is going to be open to all users, unless the OS has some control over which port is going to be allowed for which user - this is tricky to implement, not to mention that even then the max number of simultaneous sessions is restricted to the possible combinations of *addresses* and *ports* by the protocol. With UNIX domain sockets (which are basically named pipes), it's all different. The addresses are basically file paths, so there's virtually no limit on how many *sockets* you can open. Plus you have builtin file-like access controls which those internet protocols do not have. There are performance gains as well - UNIX domain sockets are much more lightweight than TCP sockets.

Of course UNIX domain sockets are not the only way to do IPC. But like I already mentioned, they are simply just more convenient.

**Comment by oakaigh** [comment]
> how is it changed with Unix sockets

No change. The proxy handles all the requests and routes the requests to the listening CloudCMD servers. 

There's currently no way to access a UNIX domain socket server from a browser that I know of. But it's supported from `curl` (https://superuser.com/a/925610), as well as a vast majority of HTTP frameworks and servers.

**Comment by coderaiser** [comment]
About security there is ability to use [Authorization](https://cloudcmd.io/#authorization) in Cloud Commander.

But what the most important is: how are you going to use Cloud Commander with UNIX sockets enabled?

**Comment by oakaigh** [comment]
As already mentioned many times, you use it behind a proxy.

> A good use case would be to run multiple instances of cloudcmd servers behind an authenticating proxy
(e.g. Jupyter server proxy https://github.com/jupyterhub/jupyter-server-proxy),
**where each instance is spawned under a different user.**

I suggest you take a close look at Jupyter server proxy and JupyterHub to understand how this works.

The use of UNIX domain sockets no longer necessitates authorization between the proxy and the server, as authorization is handled by file permissions.

**Pull request closed without merging**